daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/value/ValueSpec.scala

```diff
@@ -95,13 +114,132 @@ class ValueSpec extends FreeSpec with Matchers with Checkers with GeneratorDrive
+      "obeys order laws" in forEvery(Table("va", fooRecord, fooVariant)) { va =>
+        implicit val arb: Arbitrary[T] = va.injarb[Cid] map (va.inj(_))
+        checkLaws(SzP.order.laws[T])
+      }
+
+      "matches constructor rank" in {
+        val fooCp = shapeless.Coproduct[fooVariant.Inj[Cid]]
+        val quux = fooCp('quux ->> 42L)
+        val baz = fooCp('baz ->> 42L)
+        (fooVariant.inj(quux) ?|? fooVariant.inj(baz)) shouldBe scalaz.Ordering.LT
+      }
+
+      "preserves base order" in forEvery(Table("va", fooRecord, fooVariant)) { va =>
+        checkOrderPreserved[Cid](va, FooScope)
+      }
+    }
+
+    "for enum types" - {
+      "obeys order laws" in forAll(enumDetailsAndScopeGen, minSuccessful(20)) {
+        case (details, scope) =>
+          implicit val ord: Order[T] = Tag unsubst Value.orderInstance(scope)
+          forEvery(Table("va", details.values.toSeq: _*)) { ea =>
+            implicit val arb: Arbitrary[T] = ea.injarb[Cid] map (ea.inj(_))
+            checkLaws(SzP.order.laws[T])
+          }
+      }
+
+      "matches constructor rank" in forAll(enumDetailsAndScopeGen, minSuccessful(20)) {
+        case (details, scope) =>
+          implicit val ord: Order[T] = Tag unsubst Value.orderInstance(scope)
+          forEvery(Table("va", details.values.toSeq: _*)) { ea =>
+            implicit val arb: Arbitrary[T] = ea.injarb[Cid] map (ea.inj(_))
+            forAll(minSuccessful(20)) { (a: T, b: T) =>
+              inside((a, b)) {
+                case (ValueEnum(_, ac), ValueEnum(_, bc)) =>
+                  (a ?|? b) should ===((ea.values indexOf ac) ?|? (ea.values indexOf bc))
+              }
+            }
+          }
+      }
+
+      "preserves base order" in forAll(enumDetailsAndScopeGen, minSuccessful(20)) {
+        case (details, scope) =>
+          forEvery(Table("va", details.values.toSeq: _*)) { ea =>
+            checkOrderPreserved[Cid](ea, scope)
+          }
+      }
+    }
+  }
+
+}
+
+@SuppressWarnings(Array("org.wartremover.warts.Any"))
+object ValueSpec {
+  private val fooSpec =
+    'quux ->> VA.int64 :: 'baz ->> VA.int64 :: RNil
+  private val (_, fooRecord) = VA.record(Identifier assertFromString "abc:Foo:FooRec", fooSpec)
+  private val fooVariantId = Identifier assertFromString "abc:Foo:FooVar"
+  private val (_, fooVariant) = VA.variant(fooVariantId, fooSpec)
+  private val fooEnumId = Identifier assertFromString "abc:Foo:FooEnum"
+
+  private[this] val scopeOfEnumsGen: Gen[Map[Identifier, Seq[Name]]] =
+    Gen.mapOf(Gen.zip(idGen, Gen.nonEmptyContainerOf[Set, Name](nameGen) map (_.toSeq)))
+
+  private val enumDetailsAndScopeGen
+    : Gen[(Map[Identifier, VA.EnumAddend[Seq[Name]]], Value.LookupVariantEnum)] =
+    scopeOfEnumsGen flatMap { details =>
+      (
+        details transform ((name, members) => VA.enum(name, members)._2),
+        details.transform((_, members) => members.to[ImmArray]).lift)
+    }
+  /*
```

Comment:
must be some prototyping... should it go?

Reply:
Hmm...probably.